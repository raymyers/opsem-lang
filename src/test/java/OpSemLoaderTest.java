import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class OpSemLoaderTest {


    public static final String FULL_PROGRAM_1 = """
        rule Skip {
            ~
            Skip, s => s
        }
                
        rule Assign {
            ~
            Assign x a, s => Put s (Eval a s)
        }
                
        rule Seq {
            c1, s1 => s2
            c2, s2 => s3
            ~
            Seq c1 c2, s1 => s3
        }
                
        rule WhileTrue {
            bval b s1
            c, s1 => s2
            While b c, s2 => s3
            ~
            While b c, s1 => s3
        }
    """;

    @Test
    void generatAst()
    {
        OpSemLoader subject = new OpSemLoader(FULL_PROGRAM_1);
        assertTrue(subject.valid());

        ApprovalHelpers.verifyAsJson(subject.toProgram());
    }

    @Test
    void toLatex()
    {
        String input = """                    
            rule WhileTrue {
                bval b s1
                c, s1 => s2
                While b c, s2 => s3
                ~
                While b c, s1 => s3
            }
        """;
        OpSemLoader subject = new OpSemLoader(input);
        assertTrue(subject.valid());

        String latex = subject.toProgram().rules().get(0).toLatex();
        Approvals.verify(latex);
    }


    @Test
    void toLatexFullProgram()
    {
        OpSemLoader subject = new OpSemLoader(FULL_PROGRAM_1);
        assertTrue(subject.valid());

        String latex = subject.toProgram().toLatex();
        Approvals.verify(latex);
    }
}
