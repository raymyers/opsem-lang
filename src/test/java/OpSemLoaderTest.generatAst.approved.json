{
  "@type" : "Program",
  "rules" : [ {
    "@type" : "SemRule",
    "name" : "Skip",
    "topLines" : [ ],
    "bottomLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "Skip"
          } ]
        }, {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s"
          } ]
        } ]
      },
      "post" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s"
          } ]
        } ]
      }
    } ]
  }, {
    "@type" : "SemRule",
    "name" : "Assign",
    "topLines" : [ ],
    "bottomLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "Assign"
          }, {
            "@type" : "Var",
            "name" : "x"
          }, {
            "@type" : "Var",
            "name" : "a"
          } ]
        }, {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s"
          } ]
        } ]
      },
      "post" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "Put"
          }, {
            "@type" : "Var",
            "name" : "s"
          }, {
            "@type" : "ExprsWrap",
            "exprs" : {
              "exprs" : [ {
                "@type" : "Var",
                "name" : "Eval"
              }, {
                "@type" : "Var",
                "name" : "a"
              }, {
                "@type" : "Var",
                "name" : "s"
              } ]
            }
          } ]
        } ]
      }
    } ]
  }, {
    "@type" : "SemRule",
    "name" : "Seq",
    "topLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "c1"
          } ]
        }, {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s1"
          } ]
        } ]
      },
      "post" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s2"
          } ]
        } ]
      }
    }, {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "c2"
          } ]
        }, {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s2"
          } ]
        } ]
      },
      "post" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s3"
          } ]
        } ]
      }
    } ],
    "bottomLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "Seq"
          }, {
            "@type" : "Var",
            "name" : "c1"
          }, {
            "@type" : "Var",
            "name" : "c2"
          } ]
        }, {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s1"
          } ]
        } ]
      },
      "post" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s3"
          } ]
        } ]
      }
    } ]
  }, {
    "@type" : "SemRule",
    "name" : "WhileTrue",
    "topLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "bval"
          }, {
            "@type" : "Var",
            "name" : "b"
          }, {
            "@type" : "Var",
            "name" : "s1"
          } ]
        } ]
      },
      "post" : null
    }, {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "c"
          } ]
        }, {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s1"
          } ]
        } ]
      },
      "post" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s2"
          } ]
        } ]
      }
    }, {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "While"
          }, {
            "@type" : "Var",
            "name" : "b"
          }, {
            "@type" : "Var",
            "name" : "c"
          } ]
        }, {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s2"
          } ]
        } ]
      },
      "post" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s3"
          } ]
        } ]
      }
    } ],
    "bottomLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "While"
          }, {
            "@type" : "Var",
            "name" : "b"
          }, {
            "@type" : "Var",
            "name" : "c"
          } ]
        }, {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s1"
          } ]
        } ]
      },
      "post" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "s3"
          } ]
        } ]
      }
    } ]
  } ]
}